### PR TITLE
Assignment 2 - Zach Lawson and Tom Rohr (@Tom-Rohr)

### PR DESCRIPTION
**Pull request by zlawson-ewu**
Assignment 2 peer programed by Zach Lawson (@zlawson-ewu)  and Tom Rohr (@Tom-Rohr). Extra credit attempted with console logger. Unit Tests for all bullet points and for console logger. Did not attend to 'nameof()' bullet point, did not understand what it was asking for.

### CODE REVIEW

File: Logger.Tests/LogFactoryTests.cs [hunk 1]
```diff
@@ -1,11 +1,78 @@
...
+
+    [TestMethod]
+    public void LogFactory_CreateLogger_ReturnsLoggerIfConfigured()
+    {
+        //Arrange
+
+        //Act
+        LogFactory factory = new();
+        factory.ConfigureFileLogger(testPath);
+        BaseLogger logger = factory.CreateLogger(GetType().Name);
```
Comment (Tom-Rohr) [inline Logger.Tests/LogFactoryTests.cs:19]: I don't think our logger is writing the correct information to the file still. We should try to figure out how to print the class name from where log is being called before our final submission tonight.

**Review by Tom-Rohr — COMMENTED** [review]
-There is an existing `BaseLogger` class. It needs an **auto property** to hold the class name. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ❌
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Set `LangVersion` to 11.0 ✔
- Turn on Nullability (`Nullable`) ✔
-EXTRA CREDIT
- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ✔
Implement the factory pattern using static abstract methods on a Logger interface instead.❌

**Review by zlawson-ewu — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold the class name. This property should be set in the LogFactory using an object initializer. ✔

Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
        The current date/time ✔
        The name of the class that created the logger ✔
        The log level ✔
        The message ✔
        The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"

The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔

If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔

Inside of BaseLoggerMixins implement extension methods on BaseLogger for
        Error, ✔
        Warning, ✔
        Information, and ✔
        Debug. ✔ 
Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.

Use the nameof() operator when identifying the class name to the logger ❌
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.

Extra Credit

Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ✔

File: Logger.Tests/LogFactoryTests.cs (hunk 1 above)
Comment (zlawson-ewu) [reply Logger.Tests/LogFactoryTests.cs]: This correction has been made after receiving clarification from the instructor. The intent was to essentially hardcode (but this is fine for our purposes) the name of the class in the tests. So line 19 here now uses nameof(LogFactory) in place of GetType().Name.

File: Logger.Tests/FileLoggerTests.cs [hunk 2]
```diff
@@ -1,10 +1,30 @@
...
     {
-        
+        //Arrange
+        string date = DateTime.Now.ToString();
+        string testMessage = $"{date} {GetType().Name} {LogLevel.Information} : test";
+
+        //Act
+        LogFactory factory = new();
+        factory.ConfigureFileLogger(testPath);
+        BaseLogger logger = factory.CreateLogger(GetType().Name);
```
[the remark this answers is not in the document]
Comment (zlawson-ewu) [reply Logger.Tests/FileLoggerTests.cs]: This correction has been made after receiving clarification from the instructor. The intent was to essentially hardcode (but this is fine for our purposes) the name of the class in the tests. So line 23 here now uses nameof(FileLoggerTests) in place of GetType().Name.

**Comment by mmwoodfo** [comment]
@JLee35 @mmwoodfo exempt from doing a PR review on their own PR. The class docs were not clear and have been updated to fix the discrepancy.

File: Logger/FileLogger.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+namespace Logger;
+
+internal class FileLogger : BaseLogger
+{
+    private string? _Path;
+
+    public FileLogger(string? path)
+    {
+        _Path = path;
+    }
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:13]:
```suggestion
    private string _Path;

    public FileLogger(string path)
    {
        _Path = path ?? throw new ArgumentNullException(path);
    }
```

I would recommend not making this nullable and doing a null check in your constructor to ensure the user isn't forcing null into your constructor.

File: Logger/BaseLoggerMixins.cs [hunk 4]
```diff
@@ -1,7 +1,43 @@
...
+
+public static class BaseLoggerMixins
 {
-    public static class BaseLoggerMixins
+    public static void Error(this BaseLogger logger, string message, params object[] args)
     {
-        
+        if (logger is null)
+        {
+            throw new ArgumentNullException(nameof(logger));
```
Comment (mmwoodfo) [inline Logger/BaseLoggerMixins.cs:11]:
```suggestion
            throw new ArgumentNullException(nameof(logger));
```
I think you left a comment about not understanding the use of nameof, however, it appears that you are using it correctly here. Is there something I can clarify?

File: Logger.Tests/FileLoggerTests.cs [hunk 5]
```diff
@@ -1,10 +1,30 @@
...
     {
-        
+        //Arrange
+        string date = DateTime.Now.ToString();
+        string testMessage = $"{date} {GetType().Name} {LogLevel.Information} : test";
+
+        //Act
+        LogFactory factory = new();
+        factory.ConfigureFileLogger(testPath);
+        BaseLogger logger = factory.CreateLogger(nameof(FileLoggerTests));
```
Comment (mmwoodfo) [inline Logger.Tests/FileLoggerTests.cs:23]: Similarly, looks like you are using nameof correctly, let me know if you still have questions around this.

File: Logger/FileLogger.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+    public override void Log(LogLevel logLevel, string message)
+    {
+        StreamWriter writer = File.AppendText(_Path!);
```
Comment (mmwoodfo) [inline Logger/FileLogger.cs:17]: Consider using a `using` statement here, sounds like Mark will cover this sometime in the next few classes.

Comment (mmwoodfo) [inline Logger/FileLogger.cs:17]:
```suggestion
        StreamWriter writer = File.AppendText(_Path);
```
Avoid using this whenever you can!! If you take my above suggestions for lines 8-13, you won't need this here anymore.

If you ever find yourself typing `!` ask yourself why you're doing that, and if there is a better solution.

File: Logger/LogFactory.cs [hunk 7]
```diff
@@ -1,11 +1,24 @@
...
+    private string? _Path;
+
+    public BaseLogger CreateLogger(string className)
     {
-        public BaseLogger CreateLogger(string className)
+        if (_Path is null)
         {
-            
-            return null;
+            return null!;
```
Comment (mmwoodfo) [inline Logger/LogFactory.cs:11]:
```suggestion
            return null;
```
Same comment here, avoid using `!`, the better solution would be to change your method signature on line 7 to return a nullable `BaseLogger`

So something like this:
`public BaseLogger? CreateLogger(string className)`

File: Logger.Tests/LogFactoryTests.cs [hunk 8]
```diff
@@ -1,11 +1,78 @@
...
+    readonly string testPath = Path.Combine(Directory.GetCurrentDirectory(), "testLog.txt");
+
+    [TestMethod]
+    public void LogFactory_CreateLogger_ReturnsLoggerIfConfigured()
+    {
+        //Arrange
+
+        //Act
+        LogFactory factory = new();
+        factory.ConfigureFileLogger(testPath);
```
Comment (mmwoodfo) [inline Logger.Tests/LogFactoryTests.cs:18]: These should go into your arrange section

File: Logger/BaseLoggerMixins.cs (hunk 4 above)
Comment (zlawson-ewu) [reply Logger/BaseLoggerMixins.cs:11]: No, it's all good. Below the PR's there are collapsed line-specific conversations I resolved once I received clarification.

File: Logger.Tests/FileLoggerTests.cs (hunk 5 above)
Comment (zlawson-ewu) [reply Logger.Tests/FileLoggerTests.cs:23]: See other related response.

File: Logger/FileLogger.cs (hunk 3 above)
Comment (zlawson-ewu) [reply Logger/FileLogger.cs:13]: That is likely a better fail-safe solution. We figured that the intent was for FileLoggers to only come from LogFactory, wherein the path is null-checked.

File: Logger/FileLogger.cs (hunk 6 above)
Comment (zlawson-ewu) [reply Logger/FileLogger.cs:17]: Noted for future reference.

File: Logger.Tests/LogFactoryTests.cs (hunk 8 above)
Comment (zlawson-ewu) [reply Logger.Tests/LogFactoryTests.cs:18]: Fair enough, an oversight on our end.

File: Logger/LogFactory.cs (hunk 7 above)
Comment (zlawson-ewu) [reply Logger/LogFactory.cs:11]: Yes, that does seem to simplify things in the class, though I did have to add the operators to pieces of several tests as a result of modifying to your suggestion. Better in the tests than in the classes, I suppose.

File: Logger/FileLogger.cs (hunk 3 above)
Comment (zlawson-ewu) [reply Logger/FileLogger.cs:13]: That is likely a better fail-safe solution. We figured that the intent was for FileLoggers to only come from LogFactory, wherein the path is null-checked.

File: Logger/FileLogger.cs (hunk 6 above)
Comment (zlawson-ewu) [reply Logger/FileLogger.cs:17]: Modifying to your above suggestions has indeed removed the need for this operator.

**Pull request closed without merging**